Template/Trigger/ElementVisibilityTrigger.web.js

```diff
@@ -272,47 +257,148 @@
+                        dynamicObservedNodesForIntersection.push(element);
                     }
                 }
             });
         }
 
-        function isNodeEventTriggered(node) {
-            for (var j = 0; j < triggeredNodes.length; j++) {
-                if (node === triggeredNodes[j]) {
+        function isNodeInViewport(node) {
+            var rect = node.getBoundingClientRect();
+
+            return (
+                rect.top >= 0 &&
+                rect.left >= 0 &&
+                rect.bottom <= (windowAlias.innerHeight || documentAlias.documentElement.clientHeight) && /* or $(window).height() */
+                rect.right <= (windowAlias.innerWidth || documentAlias.documentElement.clientWidth) /* or $(window).width() */
+            );
+        }
+
+        function isDynamicNodeObservedForIntersection(node) {
+            for (var i = 0; i < dynamicObservedNodesForIntersection.length; i++) {
+                if (node === dynamicObservedNodesForIntersection[i]) {
                     return true;
                 }
             }
 
             return false;
         }
 
+        function commonTrigger(triggerEvent, percentVisible, node) {
+            var dom = TagManager.dom;
+            triggerEvent({
+                event: 'mtm.ElementVisibility',
+                'mtm.elementVisibilityPercentage': Math.round(percentVisible * 100) / 100,
+                'mtm.elementVisibilityId': dom.getElementAttribute(node, 'id'),
+                'mtm.elementVisibilityClasses': dom.getElementClassNames(node),
+                'mtm.elementVisibilityText': TagManager.utils.trim(node.innerText),
+                'mtm.elementVisibilityNodeName': node.nodeName,
+                'mtm.elementVisibilityUrl': node.href || dom.getElementAttribute(node, 'href')
+            });
+        }
+
+        function commonTriggeredNodeCheck(node) {
+            if (fireTriggerWhen === 'oncePage') {
+                blockTrigger = true;
+                if (self.scrollIndex) {
+                    TagManager.window.offScroll(self.scrollIndex);
+                }
+                if (observerIntersection) {
+                    observerIntersection.disconnect();
+                }
+                if (observerMutation) {
+                    observerMutation.disconnect();
+                }
+            } else if (onlyOncePerElement) {
+                triggeredNodes.push(node); // to avoid possible memory leaks as much as possible we add onceElement only when needed
+                if (observerIntersection) {
+                    observerIntersection.unobserve(node);
+                }
+            }
+        }
+
+        function getDynamicNodes(selectors) {
+            var dynamicNodes = [];
+            selectors.split(',').forEach(function (node) {
+                if (!documentAlias.querySelector(node)) {
+                    dynamicNodes.push(node.trim());
+                }
+            });
+
+            return dynamicNodes;
```

Comment:
do I see this right this actually returns a list of selectors not nodes?

Not 100% sure why we are doing this logic as later in `mutationObserverCallback` we query all selectors again anyway?

Reply:
Removed above function